keras_nlp/layers/relative_attention.py

```diff
@@ -0,0 +1,560 @@
+# Copyright 2022 The TensorFlow Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Keras-based relative attention layers."""
+import math
+import string
+
+import tensorflow as tf
+from tensorflow import keras
+
+from keras_nlp.api_export import keras_nlp_export
+
+_CHR_IDX = string.ascii_lowercase
+
+
+def _build_proj_equation(free_dims, bound_dims, output_dims):
+    """Builds an einsum equation for projections inside multi-head attention."""
+    input_str = ""
+    kernel_str = ""
+    output_str = ""
+    bias_axes = ""
+    letter_offset = 0
+    for i in range(free_dims):
+        char = _CHR_IDX[i + letter_offset]
+        input_str += char
+        output_str += char
+
+    letter_offset += free_dims
+    for i in range(bound_dims):
+        char = _CHR_IDX[i + letter_offset]
+        input_str += char
+        kernel_str += char
+
+    letter_offset += bound_dims
+    for i in range(output_dims):
+        char = _CHR_IDX[i + letter_offset]
+        kernel_str += char
+        output_str += char
+        bias_axes += char
+    equation = "%s,%s->%s" % (input_str, kernel_str, output_str)
+
+    return equation, bias_axes, len(output_str)
+
+
+def _get_output_shape(output_rank, known_last_dims):
+    return [None] * (output_rank - len(known_last_dims)) + list(known_last_dims)
+
+
+def _rel_shift(x, klen=-1):
+    """Performs relative shift to form the relative attention score."""
+
+    x = tf.transpose(x, perm=[2, 3, 0, 1])
+    x_size = tf.shape(x)
+
+    x = tf.reshape(x, [x_size[1], x_size[0], x_size[2], x_size[3]])
+    x = tf.slice(x, [1, 0, 0, 0], [-1, -1, -1, -1])
+    x = tf.reshape(x, [x_size[0], x_size[1] - 1, x_size[2], x_size[3]])
+    x = tf.slice(x, [0, 0, 0, 0], [-1, klen, -1, -1])
+
+    x = tf.transpose(x, perm=[2, 3, 0, 1])
+
+    return x
+
+
+@keras_nlp_export("keras_nlp.layers.MultiHeadRelativeAttention")
+class MultiHeadRelativeAttention(keras.layers.MultiHeadAttention):
+    """A multi-head attention layer with relative attention + position encoding.
+    This layer shares the same input/output projections as the common
+    `keras.layers.MultiHeadAttention` layer.
+    When it calculates attention logits, position encoding is projected to form
+    relative keys. The logits are composed by shifted relative logits and content
+    logits.
+    **Note: This layer is currently experimental.
```

Comment:
Generally, we have a blank line in between paragraphs. 

The format is as follows:
"""\<one liner explanation\>

<paragraph 1>

<paragraph 2>

Args:
arg_1: ...
arg_2: ...

Examples:
...
"""

Check any other file to get an idea of how documentation is structured.